### ISSUE TITLE
Decoupling preprocessing and training

### ISSUE THREAD

**Issue by jsimsa** [tensorflow/tensorflow#27728]
**Describe the feature and the current behavior/state.**

Currently, TensorFlow performs preprocessing and training on the same host. In situations where the data preprocessing is efficient and yet the CPU resources available on the host are not sufficient to keep up with the training workload on the accelerator, the preprocessing becomes a bottleneck. 

The proposed feature is to extend the tf.data / tf.distribute APIs to make it possible to decouple the preprocessing from training.

**Will this change the current api? How?**

Yes. In the least, the users will need to specify the set of "input" hosts that should perform the preprocessing and the set of "training" hosts to perform the training. 

The preferred solution would allow users to express their input pipeline in tf.data as if it was executing on the same host as training and through the means of tf.data / tf.distribute configuration express how it should be distributed.

**Who will benefit with this feature?**

Users that execute preprocessing intensive training jobs.

**Comment by yourtheron** [comment]
As per tf's doc, users cannot perform any transformations after using tf.strategy.experimental_distribute_dataset() to transfer a normal tf dataset into a dataset for distributed training [https://www.tensorflow.org/api_docs/python/tf/distribute/Strategy#experimental_distribute_dataset](url). Moreover, what is often in my case is that I need to perform data augmentation in each global minibatch, rather than the entire training set, of the data. 
For example in medical image segmentation, I need to randomly distort one 4D CT/MRI images in each global minibatch before feeding them into a model for forward propagation. The pesudo code is like this
```
    for ep in np.arange(0, epoch):
        # Each step crosses one minibatch
        for rawDataperGlbMB in enumerate(dsTrain):
            # Augment data from a minibatch of raw data in reshuffled dsTrain
            data = dataAugmentation(rawDataperGlbMB, ...)
	    prediction = myModel(data)

```
Therefore, when using tf.distribute.MirroredStrategy(), how to achieve that? Also, how to do reshuffling prior to each epoch?
Furthermore, if operations in my data augmentation function cannot be performed by using tf functions, but by using numpy and scipy functions (since tf functions are not sufficiently rich), can I transform rawDataperGlbMB back into numpy arrays, do the augmentations, then transform the result back into tensors, in the distributed context as I did in the non-distributed training?
Thanks.

**Comment by yourtheron** [comment]
From tf's Youtube tutorial, I became aware that tf.py_function can be used for data augmentation for each replica, yet if I write code as follows:
```
            for dataPerReplica in ds:
		tf.py_function(dataAugFuncPython, dataPerReplica, Tout)
                lossPerReplica, accPerReplic = strategy.run(trainOneReplica, args=(Tout,))
                totalLoss += strategy.reduce(tf.distribute.ReduceOp.SUM, lossPerReplica, axis=None)
                totalAcc += strategy.reduce(tf.distribute.ReduceOp.SUM, accPerReplic, axis=None)
```
where dataAugFuncPython is the name of my augmentation function in python, and trainOneReplica is the name of function for one training step. 
My question is: 
1) if dataAugFuncPython contains random variable generator, which is often the case for data augmentation, can @tf.function be decorated for the function to train the replica?
2) will the SAME augmented data be passed to all GPU's? If so, the data augmentation in this way will become meaningless. Or will data in each replica will be augmented differently automatically?

**Comment by jsimsa** [comment]
@yourtheron is better suited to be posted on Stack Overflow. Github is meant to be used for reporting bugs or requesting functionality. I am also going to close this feature request as it has been addressed by [tf.data service](https://github.com/tensorflow/community/pull/195), which was released in TF 2.3.
